example/glide_computer.cpp

```diff
@@ -101,7 +101,8 @@ using namespace si::unit_constants;
 
 template<QuantityKind QK1, QuantityKind QK2>
 constexpr Quantity auto operator/(const QK1& lhs, const QK2& rhs)
-  requires requires { lhs.common() / rhs.common(); } {
+  requires (!std::same_as<typename QK1::kind_type::base_kind, typename QK2::kind_type::base_kind>) &&
+        requires { lhs.common() / rhs.common(); } {
   return lhs.common() / rhs.common();
 }
```

Comment:
Do you think that the existence of this could hint at a problem?

Could the code benefit from dimensionless quantity kinds for the different ratios not provided by `quantity_kind` that it uses?

Reply:
Actually, I start to worry that it is the dead-end :-( What if I have `x`, `y`, `z` kinds? `length<x> * length<x>` should not be allowed.

The more I think about this I start to think that a kind should be semiregular empty type. There should be a `regular_kind` that will implement current logic (addition/subtraction for kinds and multiplication/division for scalars and quantities).

Other operations could be optionally enabled by:

```cpp
horizontal_area_kind operator*(horizontal_length_kind, horizontal_length_kind);
```

With such an approach the user would be able to define the allowed operations and their outcomes for kind-specific arithmetic. It would allow us to provide 1D, 2D, and 3D kinds or even XYZ kinds that will allow us to add X vector to Y vector and end up with XY vector which value/length will be the result of trigonometric functions usage?

What do you think about it?

Reply:
Reading your comment made me realize that the current state of the PR would leave `quantity_kind` in the right state.

> Actually, I start to worry that it is the dead-end :-( What if I have `x`, `y`, `z` kinds? `length<x> * length<x>` should not be allowed.

Why not? Surely, the strong typing has to covered. Don't care about multiplying your quantity kinds? Don't do it. Accidentally do it? Compiler error! Just like how I don't have to fear all the operators and dimensions and their inverses I don't (yet) use when using `quantity`.
> 
> The more I think about this I start to think that a kind should be semiregular empty type. There should be a `regular_kind` that will implement current logic (addition/subtraction for kinds and multiplication/division for scalars and quantities).
> 

You mean like opting into operations by incrementally marking your `kind`? As with above, I don't see the need to worry.
> Other operations could be optionally enabled by:
> 
> ```c++
> horizontal_area_kind operator*(horizontal_length_kind, horizontal_length_kind);
> ```
> 
> With such an approach the user would be able to define the allowed operations and their outcomes for kind-specific arithmetic. It would allow us to provide 1D, 2D, and 3D kinds or even XYZ kinds that will allow us to add X vector to Y vector and end up with XY vector which value/length will be the result of trigonometric functions usage?
> 
> What do you think about it?

I think the state of this PR has us covered. It provides all the inter-dimensional analysis of `quantity` through _intra_-kind operators. Then, by design (as despicted by the "user-extensible interface tests"), you can add _inter_-kind operators.

Reply:
Although you raised fair points, did you see the particular issue I actually meant?

For example, removing the piece of code we're commenting on and compiling results in:
```
glide_computer.cpp:132:18: error: no match for ‘operator/’ (operand types are ‘const velocity’ {aka ‘const units::quantity_kind<{anonymous}::velocity_vector, units::physical::si::kilometre_per_hour>’} and ‘units::quantity_kind<{anonymous}::rate_of_climb_vector, units::physical::si::metre_per_second>’)
  132 |   return polar.v / -polar.climb;
      |          ~~~~~~~ ^ ~~~~~~~~~~~~
      |                |   |
      |                |   quantity_kind<{anonymous}::rate_of_climb_vector,units::physical::si::metre_per_second>
      |                quantity_kind<{anonymous}::velocity_vector,units::physical::si::kilometre_per_hour>

glide_computer.cpp:314:103: error: no match for ‘operator/’ (operand types are ‘units::quantity_kind<{anonymous}::vertical_vector, units::physical::si::metre>’ and ‘const units::quantity_kind<{anonymous}::horizontal_vector, units::physical::si::kilometre>’)
  314 |   return distance((ground_alt + s.min_agl_height - point.alt).common() / ((ground_alt - t.finish.alt) / dist_to_finish - 1 / glide_ratio(g.polar[0])));
      |                                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~
      |                                                                                       |                 |
      |                                                                                       |                 quantity_kind<{anonymous}::horizontal_vector,units::physical::si::kilometre>
      |                                                                                       quantity_kind<{anonymous}::vertical_vector,units::physical::si::metre>
```
Here, we're mixing quantities of unrelated kinds. These requires prorgam-defined operators.

The problem I alluded to was the lack of a name, as evidenced by returning a `Dimensionless` (`lhs.common() / rhs.common()`).

Are you of the opinion that each of these should have its own meaningful name for this particular application? Perhaps if only it weren't "simplified", as mentioned in the example, but the actual, realistic application. Such that you could benefit from the stronger typing when initially writing it and under maintenance.

The operations in the errors above could, perhaps, return dimensionless quantities of kind `ratio_between_climb_and_velocity` and `ratio_of_horizontal_and_vertical_vectors`. (With better names, as better determined by the application expert. Perhaps the names in the _Simplifications:_ comment mentions some of these).

The same could be applied to the two `const duration d = (a / b).common();`. Could these, perhaps, also benefit from stronger typing (so as to not implicitly and silently mix them), if only in the real application?

Of course, it could also be fine as it is. That's your call. If that weren't the case for the real application, should we call that out in the example?

Reply:
Sorry for the late answer. I provided a few days long C++ training and when I do so I am always out from my regular life for a while :-(

Just to be clear, my comment was not about a custom operator in a `glide_computer` but about the changes to the generic operators provided by this PR.

> Don't care about multiplying your quantity kinds? Don't do it. Accidentally do it? Compiler error!

Actually with the changes of this PR `length<x> * length<x>` will compile fine, right? My comment was that it might be better to actually always work according to the customization points. So the arithmetic operation on a `quanity_kind` will only be done when it is provided for its kind.

Anyway, we can try to go with what is this PR and optimize later on based on the feedback and future use cases.

Reply:
> Actually with the changes of this PR `length<x> * length<x>` will compile fine, right? My comment was that it might be better to actually always work according to the customization points. So the arithmetic operation on a `quanity_kind` will only be done when it is provided for its kind.

That's done for all operators:
```C++
  requires requires { lhs.common() * rhs.common(); }
```